controller/engine_controller.go

```diff
@@ -1184,44 +1195,45 @@ func checkSizeBeforeRestoration(log logrus.FieldLogger, engine *longhorn.Engine,
 	return true, nil
 }
 
-func restoreBackup(log logrus.FieldLogger, engine *longhorn.Engine, rsMap map[string]*longhorn.RestoreStatus, cliAPIVersion int, ds *datastore.DataStore, engineClientProxy engineapi.EngineClientProxy) error {
+func (m *EngineMonitor) restoreBackup(engine *longhorn.Engine, rsMap map[string]*longhorn.RestoreStatus, cliAPIVersion int, engineClientProxy engineapi.EngineClientProxy) error {
 	// Get default backup target
-	backupTarget, err := ds.GetBackupTargetRO(types.DefaultBackupTargetName)
+	backupTarget, err := m.ds.GetBackupTargetRO(types.DefaultBackupTargetName)
 	if err != nil {
 		if !apierrors.IsNotFound(err) {
 			return err
 		}
 		return fmt.Errorf("cannot found the %s backup target", types.DefaultBackupTargetName)
 	}
 
-	backupTargetClient, err := getBackupTargetClient(ds, backupTarget)
+	backupTargetClient, err := getBackupTargetClient(m.ds, backupTarget)
 	if err != nil {
 		return errors.Wrapf(err, "cannot get backup target config for backup restoration of engine %v", engine.Name)
 	}
 
+	m.logger.Infof("Prepare to restore backup, backup target: %v, backup volume: %v, requested restored backup name: %v, last restored backup name: %v",
+		backupTargetClient.URL, engine.Spec.BackupVolume, engine.Spec.RequestedBackupRestore, engine.Status.LastRestoredBackup)
 	if cliAPIVersion < engineapi.CLIVersionFour {
 		// For compatible engines, `LastRestoredBackup` is required to indicate if the restore is incremental restore
-		log.Infof("Prepare to restore backup, backup target: %v, backup volume: %v, requested restored backup name: %v, last restored backup name: %v",
-			backupTargetClient.URL, engine.Spec.BackupVolume, engine.Spec.RequestedBackupRestore, engine.Status.LastRestoredBackup)
 		if err = engineClientProxy.BackupRestore(engine, backupTargetClient.URL, engine.Spec.RequestedBackupRestore, engine.Spec.BackupVolume, engine.Status.LastRestoredBackup, backupTargetClient.Credential); err != nil {
-			if extraErr := handleRestoreErrorForCompatibleEngine(log, engine, rsMap, err); extraErr != nil {
+			if extraErr := handleRestoreErrorForCompatibleEngine(m.logger, engine, rsMap, m.restoreBackoff, err); extraErr != nil {
 				return extraErr
 			}
 		}
 	} else {
 		if err = engineClientProxy.BackupRestore(engine, backupTargetClient.URL, engine.Spec.RequestedBackupRestore, engine.Spec.BackupVolume, "", backupTargetClient.Credential); err != nil {
-			log.Infof("Prepare to restore backup, backup target: %v, backup volume: %v, requested restored backup name: %v",
-				backupTargetClient.URL, engine.Spec.BackupVolume, engine.Spec.RequestedBackupRestore)
-			if extraErr := handleRestoreError(log, engine, rsMap, err); extraErr != nil {
+			if extraErr := handleRestoreError(m.logger, engine, rsMap, m.restoreBackoff, err); extraErr != nil {
 				return extraErr
 			}
 		}
 	}
+	if err == nil || !strings.Contains(err.Error(), restoreGetLockFailedMsg) {
+		m.restoreBackoff.DeleteEntry(engine.Name)
+	}
```

Comment:
NIT: Can we handle this in `handleRestoreError` or `handleRestoreErrorForCompatibleEngine`?
(The code for compatible engines will be removed in the future hence we don't need to worry about duplication)

Reply:
Here delete the backupoff entry for successfully restoring backup, and it will not get into `handleRestoreError` or `handleRestoreErrorForCompatibleEngine` if `err == nil`, so I don't think this judgement can be moved into those two functions. `!strings.Contains(err.Error(), restoreGetLockFailedMsg` check for other two skipped error types.

Reply:
Got it! Make sense.